### ISSUE TITLE
Multi-process does not work

### ISSUE THREAD

**Issue by awang** [MaterializeInc/materialize#3448]
Materialize crashes when attempting to use 2 processes.

### What version of Materialize are you using?

```
$ materialized -v
v0.3.0 (c9408416e)
```

### How did you install Materialize?

* [X] macOS release tarball

### Is the issue reproducible? If so, please provide reproduction instructions.

Process 1:
```
[21:57][awang@MacBook-Pro:dev]$ materialized --w 1 --processes 2 --process 1
materialized v0.3.0 (c9408416e) listening on 0.0.0.0:6876...
materialized encountered an internal error and crashed.

 thread: worker thread 0
message: assertion failed: input.frontier().frontier().iter().all(|t1|
                                           input_frontier.iter().any(|t2:
                                                                          &G::Timestamp|
                                                                         t2.less_equal(t1)))
   0: backtrace::capture::Backtrace::create
   1: backtrace::capture::Backtrace::new
   2: materialized::handle_panic
   3: core::ops::function::Fn::call
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
   7: timely::progress::subgraph::PerOperatorState<T>::schedule
   8: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
   9: timely::progress::subgraph::PerOperatorState<T>::schedule
  10: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
  11: timely::worker::Wrapper::step
  12: timely::worker::Worker<A>::step_or_park
  13: dataflow::server::Worker<A>::run
  14: tokio::runtime::context::enter
  15: std::sys_common::backtrace::__rust_begin_short_backtrace
  16: std::panicking::try::do_call
  17: ___rust_maybe_catch_panic
  18: core::ops::function::FnOnce::call_once{{vtable.shim}}
  19: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  20: std::sys::unix::thread::Thread::new::thread_start
  21: __pthread_start
```
Process 2:
```
[21:57][awang@MacBook-Pro:tb (master ✗)]$ materialized --w 1 --processes 2 --process 0                                                                                   
materialized v0.3.0 (c9408416e) listening on 0.0.0.0:6875...
materialized encountered an internal error and crashed.

 thread: recv thread 1
message: assertion failed: read > 0
   0: backtrace::capture::Backtrace::create
   1: backtrace::capture::Backtrace::new
   2: materialized::handle_panic
   3: core::ops::function::Fn::call
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: timely_communication::allocator::zero_copy::tcp::recv_loop
   7: std::sys_common::backtrace::__rust_begin_short_backtrace
   8: std::panicking::try::do_call
   9: ___rust_maybe_catch_panic
  10: core::ops::function::FnOnce::call_once{{vtable.shim}}
  11: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  12: std::sys::unix::thread::Thread::new::thread_start
  13: __pthread_start
```

**Comment by frankmcsherry** [comment]
I can repro something like this in differential, without materialize. Investigating!

**Comment by frankmcsherry** [comment]
My mistake, ish. I was able to reproduce a bug that appears to be fixed by https://github.com/TimelyDataflow/timely-dataflow/pull/331 where timely was potentially too eager in complaining about correctness. This does not seem to be the same issue as here, so it remains possible that it is a Materialize issue.

**Comment by frankmcsherry** [comment]
When piloting the bug around, it appeared to be an issue that the frontier in `arrange_core` went backwards, in the context of logging operators (based on the reported frontiers and timestamps). I'll dig in to this a bit more; it could be several things, including us just putting bad data into the source ("bad": with some seconds that shouldn't be used any more).

**Comment by frankmcsherry** [comment]
It appears we are building dataflows non-deterministically again, maybe? These are all indexes, and the catalog does seem to store indexes in a `HashMap` rather than a `BTree`, but I haven't figured out the lines of code that lead to the error.

```
materialize=> select id, name, count(*) from mz_dataflow_operators group by id, name having count < 2;
 id  |                                  name                                  | count 
-----+------------------------------------------------------------------------+-------
 310 | Index(s1, [Column(0), Column(1)])                                      |     1
 326 | Index(s1, [Column(0), Column(1)])                                      |     1
 228 | Index(s1, [Column(0), Column(1)])                                      |     1
 244 | Index(s1, [Column(0), Column(1)])                                      |     1
 407 | Index(s13, [Column(0), Column(1)])                                     |     1
 415 | Index(s13, [Column(0), Column(1)])                                     |     1
 228 | Index(s33, [Column(0), Column(1)])                                     |     1
 236 | Index(s33, [Column(0), Column(1)])                                     |     1
 311 | ErrIndex(s1, [Column(0), Column(1)])                                   |     1
 327 | ErrIndex(s1, [Column(0), Column(1)])                                   |     1
 229 | ErrIndex(s1, [Column(0), Column(1)])                                   |     1
 245 | ErrIndex(s1, [Column(0), Column(1)])                                   |     1
 408 | ErrIndex(s13, [Column(0), Column(1)])                                  |     1
 416 | ErrIndex(s13, [Column(0), Column(1)])                                  |     1
 229 | ErrIndex(s33, [Column(0), Column(1)])                                  |     1
 237 | ErrIndex(s33, [Column(0), Column(1)])                                  |     1
 310 | Index(s9, [Column(0), Column(1), Column(2)])                           |     1
 326 | Index(s9, [Column(0), Column(1), Column(2)])                           |     1
 236 | Index(s9, [Column(0), Column(1), Column(2)])                           |     1
 244 | Index(s9, [Column(0), Column(1), Column(2)])                           |     1
 311 | ErrIndex(s9, [Column(0), Column(1), Column(2)])                        |     1
 327 | ErrIndex(s9, [Column(0), Column(1), Column(2)])                        |     1
 237 | ErrIndex(s9, [Column(0), Column(1), Column(2)])                        |     1
 245 | ErrIndex(s9, [Column(0), Column(1), Column(2)])                        |     1
 407 | Index(s37, [Column(0), Column(1), Column(2), Column(3), Column(4)])    |     1
 415 | Index(s37, [Column(0), Column(1), Column(2), Column(3), Column(4)])    |     1
 408 | ErrIndex(s37, [Column(0), Column(1), Column(2), Column(3), Column(4)]) |     1
 416 | ErrIndex(s37, [Column(0), Column(1), Column(2), Column(3), Column(4)]) |     1
(28 rows)

materialize=> 
```

**Comment by frankmcsherry** [comment]
This is possibly fixed by #3453, which fixes an issues where we would instantiate dataflow operators in different orders in different workers, based on the vagaries of hash map enumeration order. I haven't seen it manifest again, and the erroneous counts above did go away, but that is only absence of evidence.

**Comment by awang** [comment]
build b0d78b4 works for me, closing
